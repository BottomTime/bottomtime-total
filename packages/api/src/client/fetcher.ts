import qs from 'querystring';
import { URL } from 'url';
import { ZodType, ZodTypeDef } from 'zod';

import { ErrorResponseSchema } from '../types';
import { ApiClientOptions } from './api-client-options';
import { HttpException } from './errors';

export type HttpMethod =
  | 'GET'
  | 'HEAD'
  | 'OPTIONS'
  | 'POST'
  | 'PUT'
  | 'PATCH'
  | 'DELETE';

export type SendRequestOptions = {
  body?: Record<string, unknown>;
  params?: Record<string, unknown>;
  ignoreErrors?: boolean;
};

export type ApiResponse<T> = {
  data: T;
  status: number;
  statusText: string;
};

export class Fetcher {
  constructor(private readonly options: ApiClientOptions = {}) {}

  private getFullUrl(url: string, query?: Record<string, unknown>): string {
    let fullUrl = url;

    if (this.options.baseURL) {
      fullUrl = new URL(url, this.options.baseURL).toString();
    }

    if (query) {
      fullUrl = `${fullUrl}?${qs.stringify(JSON.parse(JSON.stringify(query)))}`;
    }

    return fullUrl;
  }

  private async parseResponseBody<T>(
    res: Response,
    parser?: ZodType<T, ZodTypeDef, unknown>,
  ): Promise<T> {
    try {
      return parser ? parser.parse(await res.json()) : await res.json();
    } catch (error) {
      return {} as unknown as T;
    }
  }

  private async checkResponseForErrors(response: Response): Promise<void> {
    if (response.ok) return;

    const body = await this.parseResponseBody(response, ErrorResponseSchema);
    throw new HttpException(
      response.status,
      response.statusText,
      body.message,
      body,
    );
  }

  private async sendRequest(
    method: HttpMethod,
    url: string,
    options?: SendRequestOptions,
  ): Promise<Response> {
    const response = await fetch(this.getFullUrl(url, options?.params), {
      body: options?.body ? JSON.stringify(options.body) : undefined,
      credentials: 'same-origin',
      method,
      mode: 'cors',
      headers: {
        Accept: 'application/json',
        'Content-Type': 'application/json',
        ...(this.options.authToken
          ? { Authorization: `Bearer ${this.options.authToken}` }
          : {}),
      },
    });

    if (!options?.ignoreErrors) await this.checkResponseForErrors(response);
    return response;
  }

  async delete<T>(url: string): Promise<ApiResponse<T>> {
    const response = await this.sendRequest('DELETE', url);
    const data = await this.parseResponseBody<T>(response);

    return {
      data,
      status: response.status,
      statusText: response.statusText,
    };
  }

  async get<T>(
    url: string,
    query?: Record<string, unknown>,
    parser?: ZodType<T, ZodTypeDef, unknown>,
  ): Promise<ApiResponse<T>> {
    const response = await this.sendRequest('GET', url, { params: query });
    const data = await this.parseResponseBody<T>(response, parser);

    return {
      data,
      status: response.status,
      statusText: response.statusText,
    };
  }

  async head(url: string): Promise<number> {
    const { status } = await this.sendRequest('HEAD', url, {
      ignoreErrors: true,
    });
    return status;
  }

  async patch<T>(
    url: string,
    params?: Record<string, unknown>,
    parser?: ZodType<T, ZodType, unknown>,
  ): Promise<ApiResponse<T>> {
    const response = await this.sendRequest('PATCH', url, { body: params });
    const data = await this.parseResponseBody<T>(response, parser);

    return {
      data,
      status: response.status,
      statusText: response.statusText,
    };
  }

  async post<T>(
    url: string,
    params?: Record<string, unknown>,
    parser?: ZodType<T, ZodTypeDef, unknown>,
  ): Promise<ApiResponse<T>> {
    const response = await this.sendRequest('POST', url, { body: params });
    const data = await this.parseResponseBody<T>(response, parser);

    return {
      data,
      status: response.status,
      statusText: response.statusText,
    };
  }

  async postFormData<T>(
    url: string,
    formData: FormData,
  ): Promise<ApiResponse<T>> {
    const response = await fetch(this.getFullUrl(url), {
      body: formData,
      credentials: 'same-origin',
      method: 'POST',
      mode: 'cors',
      headers: {
        'Content-Type': 'multipart/form-data',
        ...(this.options.authToken
          ? { Authorization: `Bearer ${this.options.authToken}` }
          : {}),
      },
    });

    await this.checkResponseForErrors(response);
    const data = await this.parseResponseBody<T>(response);

    return {
      data,
      status: response.status,
      statusText: response.statusText,
    };
  }

  async put<T>(
    url: string,
    params?: Record<string, unknown>,
    parser?: ZodType<T, ZodTypeDef, unknown>,
  ): Promise<ApiResponse<T>> {
    const response = await this.sendRequest('PUT', url, { body: params });
    const data = await this.parseResponseBody<T>(response, parser);

    return {
      data,
      status: response.status,
      statusText: response.statusText,
    };
  }
}
