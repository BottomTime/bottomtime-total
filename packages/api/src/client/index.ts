export * from './errors';

export * from './user-profile';
export * from './user';

export * from './api-client-options';
export * from './fetcher';
export * from './client';
