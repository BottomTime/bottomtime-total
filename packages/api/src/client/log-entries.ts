import {
  ApiList,
  CreateOrUpdateLogEntryParamsDTO,
  CreateOrUpdateLogEntryParamsSchema,
  DiveSiteDTO,
  DiveSiteSchema,
  GetNextAvailableLogNumberResponseDTO,
  ListLogEntriesParamsDTO,
  ListLogEntriesResponseSchema,
  LogEntryDTO,
  LogEntrySchema,
  OperatorDTO,
  OperatorSchema,
} from '../types';
import { Fetcher } from './fetcher';

export class LogEntriesApiClient {
  constructor(private readonly apiClient: Fetcher) {}

  async listLogEntries(
    username: string,
    params?: ListLogEntriesParamsDTO,
  ): Promise<ApiList<LogEntryDTO>> {
    const { data: results } = await this.apiClient.get(
      `/api/users/${username}/logbook`,
      params,
      ListLogEntriesResponseSchema,
    );
    return results;
  }

  async getLogEntry(username: string, entryId: string): Promise<LogEntryDTO> {
    const { data } = await this.apiClient.get(
      `/api/users/${username}/logbook/${entryId}`,
      undefined,
      LogEntrySchema,
    );
    return data;
  }

  async createLogEntry(
    username: string,
    options: CreateOrUpdateLogEntryParamsDTO,
  ): Promise<LogEntryDTO> {
    const { data } = await this.apiClient.post(
      `/api/users/${username}/logbook`,
      options,
      LogEntrySchema,
    );
    return data;
  }

  async getNextAvailableLogNumber(username: string): Promise<number> {
    const {
      data: { logNumber },
    } = await this.apiClient.get<GetNextAvailableLogNumberResponseDTO>(
      `/api/users/${username}/logbook/nextLogEntryNumber`,
    );
    return logNumber;
  }

  async getMostRecentDiveSites(
    username: string,
    count?: number,
  ): Promise<DiveSiteDTO[]> {
    const { data } = await this.apiClient.get(
      `/api/users/${username}/logbook/recentDiveSites`,
      { count },
      DiveSiteSchema.array(),
    );
    return data;
  }

  async getMostRecentDiveOperators(
    username: string,
    count?: number,
  ): Promise<OperatorDTO[]> {
    const { data } = await this.apiClient.get(
      `/api/users/${username}/logbook/recentOperators`,
      { count },
      OperatorSchema.array(),
    );
    return data;
  }

  async updateLogEntry(
    ownerUsername: string,
    entryId: string,
    entryData: CreateOrUpdateLogEntryParamsDTO,
  ): Promise<LogEntryDTO> {
    const { data } = await this.apiClient.put(
      `/api/users/${ownerUsername}/logbook/${entryId}`,
      CreateOrUpdateLogEntryParamsSchema.parse(entryData),
      LogEntrySchema,
    );
    return data;
  }

  async deleteLogEntry(ownerUsername: string, entryId: string): Promise<void> {
    await this.apiClient.delete(
      `/api/users/${ownerUsername}/logbook/${entryId}`,
    );
  }
}
