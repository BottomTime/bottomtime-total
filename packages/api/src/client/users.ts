import {
  AdminSearchUsersParamsDTO,
  AdminSearchUsersResponseSchema,
  CreateUserParamsDTO,
  CurrentUserSchema,
  PasswordResetTokenStatus,
  ProfileDTO,
  ProfileSchema,
  SearchProfilesResponseDTO,
  SearchProfilesResponseSchema,
  SearchUserProfilesParamsDTO,
  SuccessFailResponseDTO,
  UserSchema,
  ValidateResetPasswordTokenResponseDTO,
} from '../types';
import { Fetcher } from './fetcher';
import { User } from './user';
import { UserProfile } from './user-profile';

export class UsersApiClient {
  constructor(private readonly apiClient: Fetcher) {}

  async isUsernameOrEmailAvailable(usernameOrEmail: string): Promise<boolean> {
    const status = await this.apiClient.head(
      `/api/users/${encodeURIComponent(usernameOrEmail)}`,
    );
    return status === 404;
  }

  async createUser(options: CreateUserParamsDTO): Promise<User> {
    const { data } = await this.apiClient.post(
      '/api/users',
      options,
      UserSchema,
    );
    return new User(this.apiClient, data);
  }

  async getCurrentUser(): Promise<User | null> {
    const { data: currentUser } = await this.apiClient.get(
      '/api/auth/me',
      undefined,
      CurrentUserSchema,
    );

    if (currentUser.anonymous) {
      return null;
    } else {
      return new User(this.apiClient, UserSchema.parse(currentUser));
    }
  }

  async getUser(usernameOrEmail: string): Promise<User> {
    const { data } = await this.apiClient.get(
      `/api/admin/users/${encodeURIComponent(usernameOrEmail)}`,
      undefined,
      UserSchema,
    );
    return new User(this.apiClient, data);
  }

  async getProfile(username: string): Promise<ProfileDTO> {
    const { data } = await this.apiClient.get(
      `/api/users/${username}`,
      undefined,
      ProfileSchema,
    );
    return data;
  }

  async login(usernameOrEmail: string, password: string): Promise<User> {
    const { data } = await this.apiClient.post(
      '/api/auth/login',
      {
        usernameOrEmail,
        password,
      },
      UserSchema,
    );
    return new User(this.apiClient, data);
  }

  async requestPasswordResetToken(usernameOrEmail: string): Promise<void> {
    await this.apiClient.post(
      `/api/users/${usernameOrEmail}/requestPasswordReset`,
    );
  }

  async validatePasswordResetToken(
    usernameOrEmail: string,
    token: string,
  ): Promise<PasswordResetTokenStatus> {
    const { data } =
      await this.apiClient.get<ValidateResetPasswordTokenResponseDTO>(
        `/api/users/${usernameOrEmail}/resetPassword`,
        { token },
      );

    return data.status;
  }

  async resetPasswordWithToken(
    usernameOrEmail: string,
    token: string,
    newPassword: string,
  ): Promise<boolean> {
    const { data } = await this.apiClient.post<SuccessFailResponseDTO>(
      `/api/users/${usernameOrEmail}/resetPassword`,
      {
        newPassword,
        token,
      },
    );

    return data.succeeded;
  }

  async verifyEmail(
    username: string,
    token: string,
  ): Promise<SuccessFailResponseDTO> {
    const { data } = await this.apiClient.post<SuccessFailResponseDTO>(
      `/api/users/${username}/verifyEmail`,
      { token },
    );
    return data;
  }

  async searchUsers(
    query: AdminSearchUsersParamsDTO,
  ): Promise<{ users: User[]; totalCount: number }> {
    const { data: response } = await this.apiClient.get(
      '/api/admin/users',
      query,
      AdminSearchUsersResponseSchema,
    );

    return {
      users: response.users.map((user) => new User(this.apiClient, user)),
      totalCount: response.totalCount,
    };
  }

  async searchProfiles(
    query: SearchUserProfilesParamsDTO,
  ): Promise<SearchProfilesResponseDTO> {
    const { data } = await this.apiClient.get(
      '/api/users',
      query,
      SearchProfilesResponseSchema,
    );
    return data;
  }

  wrapDTO(dto: unknown): User {
    return new User(this.apiClient, UserSchema.parse(dto));
  }

  wrapProfileDTO(dto: unknown): UserProfile {
    return new UserProfile(this.apiClient, ProfileSchema.parse(dto));
  }
}
