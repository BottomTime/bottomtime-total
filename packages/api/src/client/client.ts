import { AppMetricsDTO } from '../types';
import { AlertsApiClient } from './alerts';
import { ApiClientOptions } from './api-client-options';
import { AuthApiClient } from './auth';
import { CertificationsApiClient } from './certifications';
import { DiveSiteReviewsApiClient } from './dive-site-reviews';
import { DiveSitesApiClient } from './dive-sites';
import { Fetcher } from './fetcher';
import { FriendsApiClient } from './friends';
import { LogEntriesApiClient } from './log-entries';
import { MembershipsApiClient } from './memberships';
import { NotificationsApiClient } from './notifications';
import { OperatorReviewsApiClient } from './operator-reviews';
import { OperatorsApiClient } from './operators';
import { TanksApiClient } from './tanks';
import { UserAccountsApiClient } from './user-accounts';
import { UserProfilesApiClient } from './user-profiles';

export class ApiClient {
  private readonly client: Fetcher;

  readonly alerts: AlertsApiClient;
  readonly auth: AuthApiClient;
  readonly certifications: CertificationsApiClient;
  readonly diveSiteReviews: DiveSiteReviewsApiClient;
  readonly diveSites: DiveSitesApiClient;
  readonly friends: FriendsApiClient;
  readonly logEntries: LogEntriesApiClient;
  readonly memberships: MembershipsApiClient;
  readonly notifications: NotificationsApiClient;
  readonly operatorReviews: OperatorReviewsApiClient;
  readonly operators: OperatorsApiClient;
  readonly tanks: TanksApiClient;
  readonly userAccounts: UserAccountsApiClient;
  readonly userProfiles: UserProfilesApiClient;

  constructor(options?: ApiClientOptions) {
    this.client = options?.fetcher ?? new Fetcher(options);

    this.alerts = new AlertsApiClient(this.client);
    this.auth = new AuthApiClient(this.client);
    this.certifications = new CertificationsApiClient(this.client);
    this.diveSiteReviews = new DiveSiteReviewsApiClient(this.client);
    this.diveSites = new DiveSitesApiClient(this.client);
    this.friends = new FriendsApiClient(this.client);
    this.logEntries = new LogEntriesApiClient(this.client);
    this.memberships = new MembershipsApiClient(this.client);
    this.notifications = new NotificationsApiClient(this.client);
    this.operatorReviews = new OperatorReviewsApiClient(this.client);
    this.operators = new OperatorsApiClient(this.client);
    this.tanks = new TanksApiClient(this.client);
    this.userAccounts = new UserAccountsApiClient(this.client);
    this.userProfiles = new UserProfilesApiClient(this.client);
  }

  async getAppMetrics(): Promise<AppMetricsDTO> {
    const { data } = await this.client.get<AppMetricsDTO>('/api/metrics');
    return data;
  }
}
