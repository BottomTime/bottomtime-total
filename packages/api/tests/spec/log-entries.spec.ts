import mockFetch from 'fetch-mock-jest';

import {
  ApiList,
  CreateOrUpdateLogEntryParamsDTO,
  DepthUnit,
  DiveSiteDTO,
  ExposureSuit,
  ListLogEntriesParamsDTO,
  ListLogEntriesResponseSchema,
  LogEntryDTO,
  LogEntrySortBy,
  PressureUnit,
  SearchDiveSitesResponseSchema,
  SortOrder,
  TankMaterial,
  TemperatureUnit,
  TrimCorrectness,
  WeightCorrectness,
  WeightUnit,
} from '../../src';
import { Fetcher } from '../../src/client/fetcher';
import { LogEntriesApiClient } from '../../src/client/log-entries';
import DiveSiteTestData from '../fixtures/dive-sites-search-results.json';
import LogEntryTestData from '../fixtures/log-entries-search-results.json';
import { BasicUser } from '../fixtures/users';

const timestamp = new Date('2024-04-30T20:48:16.436Z');
const PartialTestData: LogEntryDTO = {
  id: 'bf1d4299-0c0b-47d4-bde1-d51f3573139b',
  createdAt: new Date('2024-07-23T12:09:55Z').valueOf(),
  timing: {
    entryTime: timestamp.valueOf(),
    timezone: 'Pacific/Pohnpei',
    duration: 45.5,
  },
  creator: BasicUser.profile,
};
const FullTestData: LogEntryDTO = {
  ...PartialTestData,
  conditions: {
    airTemperature: 80,
    surfaceTemperature: 78,
    bottomTemperature: 72,
    temperatureUnit: TemperatureUnit.Fahrenheit,
    chop: 2,
    current: 3,
    weather: 'Sunny',
    visibility: 4,
  },
  createdAt: new Date('2024-07-23T12:09:55Z').valueOf(),
  updatedAt: new Date('2024-07-23T12:09:55Z').valueOf(),
  site: {
    id: 'f0c5b4d4-2d1d-4b5d-8e7d-9b7a4d4b8f1d',
    name: 'The Wreck of the RMS Titanic',
    location: 'Atlantic Ocean',
  },
  timing: {
    ...PartialTestData.timing,
    bottomTime: 40.2,
  },
  logNumber: 444,
  depths: {
    averageDepth: 55.3,
    maxDepth: 92.3,
    depthUnit: DepthUnit.Feet,
  },
  equipment: {
    weight: 10,
    weightUnit: WeightUnit.Pounds,
    weightCorrectness: WeightCorrectness.Good,
    trimCorrectness: TrimCorrectness.Good,
    exposureSuit: ExposureSuit.Wetsuit5mm,
    hood: true,
    gloves: true,
    boots: true,
    camera: true,
    torch: true,
    scooter: true,
  },
  notes: 'Sick shipwreck!',
  air: [
    {
      name: 'lean photographer',
      material: TankMaterial.Aluminum,
      workingPressure: 300,
      volume: 4,
      count: 1,
      startPressure: 227.7898846170865,
      endPressure: 69.807624156354,
      pressureUnit: PressureUnit.Bar,
      o2Percent: 27.6,
    },
  ],
  tags: ['wreck', 'deep', 'cold'],
};

describe('Log entries API client', () => {
  let fetcher: Fetcher;
  let client: LogEntriesApiClient;
  let logEntryData: ApiList<LogEntryDTO>;
  let diveSiteData: ApiList<DiveSiteDTO>;

  beforeAll(() => {
    fetcher = new Fetcher();
    client = new LogEntriesApiClient(fetcher);
    logEntryData = ListLogEntriesResponseSchema.parse(LogEntryTestData);
    diveSiteData = SearchDiveSitesResponseSchema.parse(DiveSiteTestData);
  });

  afterEach(() => {
    mockFetch.restore();
  });

  it('will return a list of log entries', async () => {
    const username = 'greg';
    const params: ListLogEntriesParamsDTO = {
      query: 'sam',
      startDate: new Date('2021-01-01').valueOf(),
      endDate: new Date('2021-12-31').valueOf(),
      limit: 800,
      skip: 20,
      sortBy: LogEntrySortBy.EntryTime,
      sortOrder: SortOrder.Ascending,
    };
    mockFetch.get(
      {
        url: `/api/users/${username}/logbook`,
        query: {
          ...params,
          startDate: new Date(params.startDate!).valueOf(),
          endDate: new Date(params.endDate!).valueOf(),
          skip: params.skip,
          limit: params.limit,
        },
      },
      {
        status: 200,
        body: logEntryData,
      },
    );

    const result = await client.listLogEntries(username, params);

    expect(result.totalCount).toBe(logEntryData.totalCount);

    result.data.forEach((entry, index) => {
      expect(entry).toEqual(logEntryData.data[index]);
    });
  });

  it('will retrieve a single log entry', async () => {
    const entryData = logEntryData.data[0];
    mockFetch.get(
      {
        url: `/api/users/${entryData.creator.username}/logbook/${entryData.id}`,
      },
      {
        status: 200,
        body: entryData,
      },
    );

    const entry = await client.getLogEntry(
      entryData.creator.username,
      entryData.id,
    );

    expect(mockFetch.done()).toBe(true);
    expect(entry).toEqual(entryData);
  });

  it('will create a new log entry without a dive site', async () => {
    const options: CreateOrUpdateLogEntryParamsDTO = {
      timing: {
        duration: 50.5,
        entryTime: new Date('2024-04-30T20:48:16').valueOf(),
        timezone: 'Pacific/Pohnpei',
        bottomTime: 50.2,
      },
      logNumber: 555,
      depths: {
        maxDepth: 95.3,
        depthUnit: DepthUnit.Feet,
      },
      notes: 'Awesome dive!',
      air: [
        {
          count: 1,
          endPressure: 500,
          material: TankMaterial.Aluminum,
          name: 'AL80',
          pressureUnit: PressureUnit.PSI,
          startPressure: 3000,
          volume: 11.1,
          workingPressure: 207,
          o2Percent: 32,
        },
      ],
    };
    const expected: LogEntryDTO = {
      ...options,
      site: undefined,
      operator: undefined,
      creator: BasicUser.profile,
      createdAt: new Date('2024-04-30T20:48:16').valueOf(),
      id: '62389e6e-0332-4288-9d87-9bd94ba830da',
    };
    mockFetch.post(`/api/users/${BasicUser.username}/logbook`, {
      status: 201,
      body: expected,
    });

    const entry = await client.createLogEntry(BasicUser.username, options);

    expect(mockFetch.done()).toBe(true);
    expect(entry).toEqual(expected);
  });

  it('will return the next available log number', async () => {
    const username = 'greg';
    const logNumber = 77;
    mockFetch.get(`/api/users/${username}/logbook/nextLogEntryNumber`, {
      status: 200,
      body: { logNumber },
    });

    const result = await client.getNextAvailableLogNumber(username);

    expect(mockFetch.done()).toBe(true);
    expect(result).toBe(logNumber);
  });

  it('will request most recently logged dive sites', async () => {
    const username = 'carl';
    mockFetch.get(
      {
        url: `/api/users/${username}/logbook/recentDiveSites?count=15`,
        // query: { count: 15 },
      },
      {
        status: 200,
        body: diveSiteData.data.slice(0, 8),
      },
    );

    const result = await client.getMostRecentDiveSites(username, 15);

    expect(mockFetch.done()).toBe(true);
    expect(result.length).toBe(8);
    expect(result.map((site) => ({ id: site.id, name: site.name }))).toEqual(
      diveSiteData.data
        .slice(0, 8)
        .map((site) => ({ id: site.id, name: site.name })),
    );
  });

  it('will update an existing log entry', async () => {
    const options: CreateOrUpdateLogEntryParamsDTO = {
      timing: {
        duration: 50.5,
        entryTime: new Date('2024-04-30T20:48:16').valueOf(),
        timezone: 'Pacific/Pohnpei',
        bottomTime: 50.2,
      },
      conditions: { ...FullTestData.conditions },
      logNumber: 555,
      depths: {
        ...FullTestData.depths,
        maxDepth: 95.3,
        depthUnit: DepthUnit.Feet,
      },
      notes: 'Awesome dive!',
      equipment: {
        ...FullTestData.equipment,
        weight: 5.2,
        weightUnit: WeightUnit.Pounds,
      },
      air: [
        {
          name: 'robust spokesman',
          material: TankMaterial.Steel,
          workingPressure: 207,
          volume: 4,
          count: 1,
          startPressure: 213.0432935175486,
          endPressure: 62.67623910983093,
          pressureUnit: PressureUnit.Bar,
          o2Percent: 21.6,
        },
      ],
      tags: FullTestData.tags,
      site: FullTestData.site!.id,
    };

    mockFetch.put(
      {
        url: `/api/users/${FullTestData.creator.username}/logbook/${FullTestData.id}`,
        body: options,
      },
      {
        status: 200,
        body: FullTestData,
      },
    );

    await client.updateLogEntry(
      FullTestData.creator.username,
      FullTestData.id,
      options,
    );
    expect(mockFetch.done()).toBe(true);
  });

  it('will delete a log entry', async () => {
    const username = 'greg';
    const entryId = '83edf350-d0c3-4fd4-a2d3-03018aff6f46';
    mockFetch.delete(`/api/users/${username}/logbook/${entryId}`, 204);

    await client.deleteLogEntry(username, entryId);

    expect(mockFetch.done()).toBe(true);
  });
});
