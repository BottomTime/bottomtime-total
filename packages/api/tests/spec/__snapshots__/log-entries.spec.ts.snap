// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Log entries API client will list most recent dive sites 1`] = `
[
  {
    "averageDifficulty": 1.3,
    "averageRating": 4.4,
    "createdOn": 1678720808407,
    "creator": {
      "accountTier": 0,
      "avatar": "https://cloudflare-ipfs.com/ipfs/Qmd3W5DuhgHirLHGVixi6V76LhCkZUz6pnFt5AJBiyvHye/avatar/1189.jpg",
      "location": "Cheyenne, VT, NR",
      "logBookSharing": "private",
      "memberSince": 1550661767372,
      "name": "Ford Kris",
      "userId": "179bebf1-cc41-48cf-88aa-0f886c01e74f",
      "username": "Ford_Kris",
    },
    "directions": "Civitas trans nobis conventus compello aro. Carcer villa vulariter apparatus benevolentia. Tergeo tabula terminatio tres magni atavus adicio certus.",
    "freeToDive": false,
    "id": "af6294af-89e4-49ea-80af-967de0f336cd",
    "location": "Maximilliaborough, WA, PS",
    "name": "agreeable, private eardrum",
    "shoreAccess": true,
  },
  {
    "averageDifficulty": 4.9,
    "averageRating": 2,
    "createdOn": 1589163161228,
    "creator": {
      "accountTier": 0,
      "avatar": "https://cloudflare-ipfs.com/ipfs/Qmd3W5DuhgHirLHGVixi6V76LhCkZUz6pnFt5AJBiyvHye/avatar/1158.jpg",
      "location": "Henderson, NE, CG",
      "logBookSharing": "private",
      "memberSince": 1421637069831,
      "name": "Kaelyn Muller",
      "userId": "bc1d6cad-6189-4f44-a9f1-0f095b6cd0ff",
      "username": "Kaelyn.Muller80",
    },
    "depth": {
      "depth": 104.7,
      "unit": "ft",
    },
    "directions": "Cometes tempora officiis commodi taceo somnus audeo odit ars cupiditate. Teres attonbitus amplitudo autus titulus ventus. Vicissitudo suggero tamquam tunc sopor adficio capitulus.",
    "freeToDive": false,
    "gps": {
      "lat": -61.5161,
      "lon": 111.4163,
    },
    "id": "2358f771-1bda-4792-895a-15ab916ab569",
    "location": "Fort Morton, NH, KG",
    "name": "authentic, key crate",
    "shoreAccess": false,
  },
]
`;

exports[`Log entries API client will list most recent operators 1`] = `
[
  {
    "active": true,
    "address": "329 Welland Ave, St. Catharines, ON L2R 2R2, Canada",
    "createdAt": 1739204161841,
    "description": "Dan’s was purchased by Matt Mandziuk who has had a big set of shoes to fill as Dan backs out slowly into retirement. We have enjoyed the new concepts Matt continues to bring to the store, as well as fresh and fun ideas. There are always many new and exciting changes with Matt at the helm. While most things have remained the same, the main goals have been to get things running more efficiently and to make the sport more accessible to everyone involved. One of the most successful changes within the brand was offering in-store financing, renovating the layout of the store, as well as the addition of a number of great new dive trips and charters. Our goal as a shop is to keep expanding, growing and improving at every turn.
Matt’s vision for the Dan’s brand has also expanded into the creation of Divers Edge – our technical dive training arm. With a solid and proven foundation of recreational diving, the Divers Edge brand aims to further technical diving and exploration with some exciting new projects and trips specific to technical divers.

2024 is going to be our 50th anniversary of Dan’s Dive Shop. To hit the Forty year mark in any business is a huge achievement – let alone the dive business – is no easy task. We couldn’t have done it without you – our customers and students. It’s been a great ride full of learning, improvement, more learning and lots of bubbles!

Thank you for your continued support of the DDS brand and we look forward to adding more to our history!

See you in the water,
The DDS family.",
    "email": "admin@dansdiveshop.ca",
    "gps": {
      "lat": 43.1708282,
      "lon": -79.2268426,
    },
    "id": "0194f0a5-7131-7bb9-95d2-a7f833eff151",
    "logo": "/api/operators/dans-dive-shop/logo",
    "name": "Dan's Dive Shop",
    "owner": {
      "accountTier": 200,
      "avatar": "/api/users/Chris/avatar",
      "location": "Cambridge, ON",
      "logBookSharing": "private",
      "memberSince": 1713361114242,
      "name": "Chris Carleton",
      "userId": "cb12fb5b-f1b3-499f-a6ae-dfb467d3d2d1",
      "username": "Chris",
    },
    "phone": "(800) 268-3267",
    "slug": "dans-dive-shop",
    "socials": {
      "facebook": "dansdiveshop",
      "instagram": "dansdiveshop",
      "tiktok": "dansdiveshop",
      "twitter": "dansdiveshop",
      "youtube": "dansdiveshop",
    },
    "updatedAt": 1739206823938,
    "verificationStatus": "unverified",
    "website": "https://dansdiveshop.ca/",
  },
]
`;

exports[`Log entries API client will retrieve dive site review 1`] = `
{
  "comments": "I love diving here. Warm water and good vis.",
  "createdOn": 1739559310897,
  "creator": {
    "accountTier": 200,
    "avatar": "/api/users/Chris/avatar",
    "location": "Cambridge, ON",
    "logBookSharing": "private",
    "memberSince": 1713361114242,
    "name": "Chris Carleton",
    "userId": "cb12fb5b-f1b3-499f-a6ae-dfb467d3d2d1",
    "username": "Chris",
  },
  "difficulty": 0.9,
  "id": "019505d0-9631-799c-951f-04f0bb2f2c13",
  "rating": 4.7,
  "updatedOn": 1739559310897,
}
`;

exports[`Log entries API client will retrieve log entry data samples 1`] = `
[
  {
    "depth": 50,
    "gps": {
      "lat": 43.1708282,
      "lng": -79.2268426,
    },
    "offset": 1000,
  },
  {
    "depth": 56,
    "gps": {
      "lat": 43.1708282,
      "lng": -79.2268426,
    },
    "offset": 2000,
  },
]
`;

exports[`Log entries API client will retrieve operator review 1`] = `
{
  "comments": "Sick dive shop, great staff!",
  "createdAt": 1739560608679,
  "creator": {
    "accountTier": 200,
    "avatar": "/api/users/Chris/avatar",
    "location": "Cambridge, ON",
    "logBookSharing": "private",
    "memberSince": 1713361114242,
    "name": "Chris Carleton",
    "userId": "cb12fb5b-f1b3-499f-a6ae-dfb467d3d2d1",
    "username": "Chris",
  },
  "id": "019505e4-63a7-7888-91a3-854ea65487a2",
  "rating": 4.1,
  "updatedAt": 1739560608679,
}
`;

exports[`Log entries API client will review dive site 1`] = `
{
  "comments": "Updated review!",
  "createdOn": 1739559310897,
  "creator": {
    "accountTier": 200,
    "avatar": "/api/users/Chris/avatar",
    "location": "Cambridge, ON",
    "logBookSharing": "private",
    "memberSince": 1713361114242,
    "name": "Chris Carleton",
    "userId": "cb12fb5b-f1b3-499f-a6ae-dfb467d3d2d1",
    "username": "Chris",
  },
  "difficulty": 2.2,
  "id": "019505d0-9631-799c-951f-04f0bb2f2c13",
  "rating": 3.8,
  "updatedOn": 1739559310897,
}
`;

exports[`Log entries API client will review operator 1`] = `
{
  "comments": "Sick dive shop, great staff!",
  "createdAt": 1739560608679,
  "creator": {
    "accountTier": 200,
    "avatar": "/api/users/Chris/avatar",
    "location": "Cambridge, ON",
    "logBookSharing": "private",
    "memberSince": 1713361114242,
    "name": "Chris Carleton",
    "userId": "cb12fb5b-f1b3-499f-a6ae-dfb467d3d2d1",
    "username": "Chris",
  },
  "id": "019505e4-63a7-7888-91a3-854ea65487a2",
  "rating": 4.1,
  "updatedAt": 1739560608679,
}
`;
